GUACAMOLE-38: Implement on-demand connections

necouchman opened the pull request:
This is an extension that implements the on-demand (ad-hoc, "QuickConnect") feature, which places a connection bar on the home page and allows users to type in a URI to establish a connection.  Everything is in-memory, so connections do not persist after logout, and there are no records kept of the connections.

The original JIRA issue was closed as a duplicate of parameter prompting, but reopened it as I think they're different enough, and went ahead and wrote it, anyway, and figured I'd see if there was any interest in merging it into the code.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/resources/controllers/quickconnectController.js:
```diff
@@ -0,0 +1,186 @@
+         * [15] - Parameters
+         * [17] - JS Route
+         */
+        var regexURL = /^(((rdp|ssh|telnet|vnc)?):\/)?\/?((.*?)(:(.*?)|)@)?([^:\/\s]+)(:([^\/]*))?((\/\w+\/)*)([-\w.\/]+[^#?\s]*)?(\?([^#]*))?(#(.*))?$/g;
```
Whoa.

necouchman replied:
Ha.  Yeah.  If you have suggestions for better ways to parse out a URI, I'm open.  That regex right there is the result of a combo of Google searches and working with the regex on regex101.com.

necouchman replied:
Obviously it would be better if the available protocols there were parsed from a REST API check of supported protocols instead of hard-coded.

mike-jumper replied:
On a similar note, any reason why you chose to handle this client-side vs. server-side? I'm not sure whether one is better than the other at the moment, but I am curious. Might be worth considering.

necouchman replied:
I think the reason I did it that way was because I was trying to leverage as much as possible of the existing REST endpoints for creating/saving the connections, which seems to be easier if the connection attributes and parameters are parsed out on the client-side and then fed to the existing API.

I suppose I could leverage the extension-specific REST API and feed the entire connection string to a REST endpoint and then parse it out there.  Or is there some other way I'm not thinking about that already exists to get the data to the server-side for parsing by the extension?

necouchman replied:
> I'm not sure whether one is better than the other at the moment, but I am curious.

Well, since Java already has a URI class and is capable of parsing out the URI without the...colorful...regex, I think I might be leaning toward redoing it server-side.  JavaScript does have some basic parsing available through the DOM, and some external libraries available for it, but it doesn't seem nearly as robust as the Java library.

necouchman commented:
@mike-jumper: Is this something you're interested in having in the Guacamole client (with changes), or is this something that seems out-of-scope to you?  I'm happy to close out this PR and maintain this extension myself - I find it very useful, but that may just be me.  If there's not interest in having it as part of the guacamole-client codebase I'll be happy to close it out so that it doesn't distract from other PRs.

mike-jumper commented:
It seems in-scope. I don't think it would make sense for this to be the default behavior of the webapp, but I definitely think it will be a welcome extension.

Have things changed in your approach since your initial push? I see some new commits have appeared.

necouchman commented:
Cool, I'll continue to push forward with this.

I've made some progress, but nothing worth any more review at this point - will ping back once I finish it up.  I just wanted to make sure this was worth the continued review before I pushed on.

Thanks!

necouchman commented:
@mike-jumper: Okay, I think this is ready for a review, now - the parsing has been implemented server-side, and seems to be functional.  Let me know what needs to be tweaked.

necouchman commented:
@mike-jumper Ping, when you get a chance to review.  Thx...

necouchman commented:
@mike-jumper <Gentle Nudge>

Or, @jmuehlner, @ceharris, @flangelo - anyone else care to review/comment?

ceharris commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java:
```diff
@@ -0,0 +1,158 @@
+     *     If an error is encountered adding the object to the
+     *     directory.
+     */
+    public String put(QuickConnection object) throws GuacamoleException {
```
Replace `object` with `connection`?

necouchman replied:
Yep, you/re right - this should be fixed, now.

ceharris commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnection.java:
```diff
@@ -0,0 +1,181 @@
+    @Override
+    public GuacamoleTunnel connect(GuacamoleClientInformation info)
+            throws GuacamoleException {
+
```
This is not a comment about the quick connect extension so much as a comment about whether there is a generalization here that should be abstracted. There's a body of about 40 lines of code here that has nothing to do with quick connect. Isn't this behavior (using the environment to get the proxy config and figure out what's needed to create an appropriate Guacamole socket) something that should be either pulled up into the abstract base type or otherwise moved into a reusable component?

necouchman replied:
Yes, you're probably right, and it's probably already implemented in some other connection.  Let me see if there's some other class that I should be `extend`ing or `implement`ing that already has this rather than having it in here.

necouchman replied:
So, I think when I originally wrote this I copied the `org.apache.guacamole.net.auth.simple.SimpleConnection` class over to this QuickConnection class, instead of just extending it.  I've refactored it to just extend the SimpleConnection object and override where I need to instead of duplicaitng the code.  Thanks for pointing that out!

necouchman commented:
Thanks @ceharris - I've tweaked the couple of items you pointed out there!  Let me know if anything else needs attention.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java:
```diff
@@ -0,0 +1,158 @@
+     *     ID to get used when adding connections.
+     */
+    private Integer getNextConnectionID() {
+        return CONNECTION_ID++;
```
This is not threadsafe. If multiple calls to `add()` overlap, this value will not be updated as expected. You will need to generate these IDs in a threadsafe manner, or switch to a mechanism that guarantees uniqueness without depending on synchronous updates ([`UUID.randomUUID()`](https://docs.oracle.com/javase/7/docs/api/java/util/UUID.html#randomUUID()) for example).

ceharris replied:
You want to be careful with `UUID.randomUUID()`. The underlying implementation uses `SecureRandom` which can block indefinitely waiting for sufficient entropy (e.g. in the kernel implementation of `/dev/random`). On a system with low activity this can be significant concern. 

`AtomicInteger.getAndIncrement` will give the same semantics as your use of the post-increment operator with thread safety. Performance is generally better than what you can do with your own `ReentrantLock` or (not recommended) an approach using `synchronized`.

mike-jumper replied:
Not persisting connections at all, and instead dynamically generating the returned `Connection` within `get()` (using the URL as the identifier) would be another option, assuming that persisting the connections in memory is not an intended feature.

necouchman replied:
Went the AtomicInteger route - having the connections persist in-memory is (to me) a desired functionality, so that accomplishes thread-safeness without losing that.  Also cleaned up the code (like my oddly-named all-caps CONNECTION_ID) a little.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java:
```diff
@@ -0,0 +1,158 @@
+    @Override
+    public void add(Connection object) throws GuacamoleException {
+
+        put(new QuickConnection(object));
```
Out of curiosity, is it an intended feature that these connections persist in memory until the end of the session? Or is that a side effect of the implementation, with the only goal being to provide some way to connect in an ad hoc manner?

necouchman replied:
It was probably initially a side-effect, but one that I liked so I made it intentional.  How's that?! :-)

Obviously the main goal is to connect in an ad-hoc manner; however, I think it's also useful to be able to quickly retrieve those connections in the same session - i.e., I get kicked out of my session due to a network issue and want to quickly reconnect without typing the entire connection string, again.

Also, in one iteration of the implementation I had a little box where it looked for other available connection directories, and, if you had permission, allowed you to use the ad-hoc box to create a connection within the directory.  I don't think I kept it, here - needs a little more work - but it might be something that gets tacked (back) on in the future.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java:
```diff
@@ -0,0 +1,158 @@
+     *     An Integer representing the next available connection
+     *     ID to get used when adding connections.
+     */
+    private Integer getNextConnectionID() {
```
Really should use `int` unless there is a reason a wrapper class like `Integer` is necessary.

necouchman replied:
Switched to `int`.  I think the reason I went `Integer` is because there were a couple of places where it gets converted to a `String`.  I've replaced those with `Integer.toString()` calls with the primitive, so accomplishes the same thing.

necouchman commented:
Okay, more tweaks - let me know what else needs to be cleaned up!

necouchman commented:
@mike-jumper These last updates look okay?  Anything else I need to work on?

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+            String[] authenticators = userInfo.split(":");
+            if (authenticators.length > 0 && authenticators[0] != null)
+                username = authenticators[0];
+            if (authenticators.length > 1 && authenticators[1] != null)
```
What if `authenticators` is something like "myusername:myPassword:That:Has:Colons"? Rather than manually parse everything with `split()`, I suggest defining (and pre-compiling) regular expressions. It would avoid recompiling the pattern each time `split()` is invoked, and may be more robust.

Alternatively, though I haven't looked into this, there may be a better class for parsing URIs which handles all of this. It doesn't look like anything we're doing here is non-standard with respect to URI format.

necouchman replied:
I haven't found a class that parses the username and password into separate components - even the URI and URL classes that have methods for grabbing the "user-info" part of the string (including both username and password) don't seem to have methods for splitting it out to individual components.  I'll work on the regex route.

necouchman replied:
Okay, implemented regex and Pattern/Matcher for this.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+        }
+        String protocol = qcUri.getScheme();
+        String host = qcUri.getHost();
+        Integer port = qcUri.getPort();
```
You should avoid using wrapper classes like this unnecessarily.

necouchman replied:
Switched to int.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/resources/templates/quickconnectField.html:
```diff
@@ -0,0 +1,9 @@
+<div class="header" ng-controller="quickconnectController">
+    <div class="quickconnect-container">
+        <form ng-submit="quickConnect()">
+            <input type=text class="quickconnect-field" placeholder="Enter Connection URI" ng-model="uri" />
```
Strings which will be displayed to the user need to be defined as translation strings, rather than hard-coded in the HTML.

necouchman replied:
Done.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/resources/styles/quickconnect.css:
```diff
@@ -0,0 +1,48 @@
+    box-sizing: border-box;
+}
+
+.quickconnect-field input[type="submit"] {
```
Is there an `<input type="submit" ...>`? I don't see one in the HTML.

necouchman replied:
Nope.  Removed.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/rest/QuickConnectREST.java:
```diff
@@ -0,0 +1,102 @@
+    public QuickConnectREST(QuickConnectUserContext userContext)
+            throws GuacamoleException {
+        this.userContext = userContext;
+        this.directory = (QuickConnectDirectory)this.userContext.getConnectionDirectory();
```
If usages of `QuickConnectUserContext` must rely on `getConnectionDirectory()` returning a `QuickConnectDirectory`, it would be better to avoid requiring typecasting and explicitly declare that as the return type.

necouchman replied:
Fixed.  I think I was originally avoiding implementing my own connection directory and trying to use one of the provided ones, and that didn't pan out and I forgot to go back and fix stuff.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/rest/QuickConnectREST.java:
```diff
@@ -0,0 +1,102 @@
+    public String create(@FormParam("uri") String uri) 
+            throws GuacamoleException {
+
+        if (directory == null)
```
Will this ever be the case?

necouchman replied:
Certainly should not.  Removed.

mike-jumper commented on extensions/guacamole-auth-quickconnect/pom.xml:
```diff
@@ -0,0 +1,209 @@
+        <!-- Jersey - JAX-RS Implementation -->
+        <dependency>
+            <groupId>com.sun.jersey</groupId>
+            <artifactId>jersey-server</artifactId>
```
You actually don't need to pull in `jersey-server` just for JAX-RS annotations. You can pull just the API containing the annotations, and declare the scope as "provided" so it doesn't get bundled:

    <!-- JAX-RS annotations -->
    <dependency>
        <groupId>javax.ws.rs</groupId>
        <artifactId>jsr311-api</artifactId>
        <version>1.1.1</version>
        <scope>provided</scope>
    </dependency>

necouchman replied:
Good to know.  Changed.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectAuthenticationProvider.java:
```diff
@@ -0,0 +1,91 @@
+
+        String username = credentials.getUsername();
+        if (username == null || username.isEmpty())
+            throw new GuacamoleInvalidCredentialsException("You must login.", CredentialsInfo.USERNAME_PASSWORD);
```
If this extension will not perform authentication, why does it request a username and password?

necouchman replied:
Good question, to which I do not have an answer.  Removed.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectAuthenticationProvider.java:
```diff
@@ -0,0 +1,91 @@
+ * process so that the QuickConnect functionality can be initialized and be used
+ * throughout the web client.
+ */
+public class QuickConnectAuthenticationProvider extends SimpleAuthenticationProvider {
```
As you have your own underlying `UserContext` implementation (a good thing), you should not be extending `SimpleAuthenticationProvider` here.

necouchman replied:
...even though doing so greatly reduces the amount of code I have to write to overwrite methods that are doing exactly the same thing as `SimpleAuthenticationProvider` - like `shutdown()` and `decorate()` and `redecorate()` and probably a couple of others that I would basically just be copying/pasting code out of `SimpleAuthenticationProvider` to duplicate that functionality?

mike-jumper replied:
I get where you're coming from, and reducing duplication is good, but what you need is an actual base implementation of `AuthenticationProvider` which guarantees the default behavior you're looking for. `SimpleAuthenticationProvider` is not guaranteed or intended to provide this, and is only defined with respect to `getAuthorizedConfigurations()`.

A base, empty implementation of `AuthenticationProvider` would probably be a good thing (similar to `AbstractUser` and others). If you're looking to avoid duplication, perhaps we should:

* Write an `AbstractAuthenticationProvider` (perhaps also `AbstractUserContext` if it helps)
* Refactor `SimpleAuthenticationProvider` to build off that if appropriate.
* Avoid using `SimpleAuthenticationProvider` here

necouchman replied:
Ah, okay, that makes sense.  I guess I'll just switch it over to `implements AuthenticationProvider` for now - I think it would probably make sense at some point in the future to create the `AbstractAuthenticationProvider` and `AbstractUserContext` implementations, but it seems like that might be more appropriate in a separate change.

necouchman replied:
Implemented.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectConnectionGroup.java:
```diff
@@ -0,0 +1,113 @@
+ * for temporarily storing the QuickConnections created by
+ * users.
+ */
+class QuickConnectConnectionGroup extends AbstractConnectionGroup {
```
Not necessarily an issue, but to confirm: is this intentionally package-private?

necouchman replied:
Well, it's never expected to be used outside this package, but neither are a bunch of the other classes, so I should probably be consistent about it...

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+public class QCParser {
+
+    /**
+     * The default protocol to parse to if one is undefined.
```
Cleaned this up.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+
+public class QCParser {
```
Documented.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * The default port to use if one is not defined.
+     */
+    public static final Integer DEFAULT_URI_PORT = 22;
```
Removed.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+
+        if (userInfo != null && !userInfo.equals("")) {
+            String[] authenticators = userInfo.split(":");
+            if (authenticators.length > 0 && authenticators[0] != null)
```
Should be cleaned up, now, with move to regex.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,169 @@
+            paramList = Arrays.asList(query.split("&"));
+
+        if (userInfo != null && !userInfo.equals("")) {
+            String[] authenticators = userInfo.split(":");
```
Switched to RegEx.

necouchman commented:
Okay, next round completed.  What's next? :-)

necouchman commented:
> What's next? :-)

Refactoring this based on the recent changes:
- `AbstractUserContext` and `AbstractAuthenticationModule` implementatoins
- Deprecation of `Simple*Directory` classes in favor of `SimpleDirectory<*>`

necouchman commented:
Okay @mike-jumper and @ceharris (and whomever else wants to review it!), this has been rebased onto the master and reworked with the recent `AbstractUserContext` and `AbsractAuthenticationProvider` classes, and also the removal of the deprecated `Simple*Directory` classes.  Net reduction of 150 lines of code :-).  Let me know what else needs to be cleaned up, here.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectAuthenticationProviderModule.java:
```diff
@@ -0,0 +1,81 @@
+     * @throws GuacamoleException
+     *     If an error occurs while retrieving the server environment.
+     */
+    public QuickConnectAuthenticationProviderModule(AuthenticationProvider authProvider)
```
Actually, I also don't see `@Inject` anywhere. Is Guice being used here at all? Perhaps this is unneeded.

necouchman replied:
Removed the whole class.

necouchman commented on extensions/guacamole-auth-quickconnect/src/licenses/NOTICE:
```diff
@@ -0,0 +1,5 @@
+Apache Guacamole
+Copyright 2017 The Apache Software Foundation
```
Just reliving the glory days.  Fixed.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectAuthenticationProvider.java:
```diff
@@ -0,0 +1,56 @@
+
+        if (userContext == null)
+            userContext = new QuickConnectUserContext(this,
+                    authenticatedUser.getIdentifier());
```
Should be good, now.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectUserContext.java:
```diff
@@ -0,0 +1,149 @@
+        // Initialize each of the directories associated with the userContext.
+        this.userDirectory = new SimpleDirectory<User>(self);
+        this.connectionDirectory = new QuickConnectDirectory(this.rootGroup);
+        this.connectionGroupDirectory = new SimpleDirectory<ConnectionGroup>(Collections.singleton(this.rootGroup));
```
Well we should probably not bother doing that, then.  Removed.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnection.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Number of active connections.
+     */
+    private int activeConnections;
```
If by "this" you mean `this` as in the entire class, then, yes.  Removed this and just using `SimpleConnection` class.  I was thinking about doing something with active connections in the QuickConnect extension once upon a time, but it doesn't seem like it has much value.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/resources/controllers/quickconnectController.js:
```diff
@@ -0,0 +1,75 @@
+        }
+    };
+
+    $scope.uri = null;
```
Documented.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,187 @@
+     *
+     * @return
+     *     The String value of the name that is generated, or
+     *     null if no config is provided.
```
Reworded.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,187 @@
+
+        String name = "";
+
+        if (protocol != null && !protocol.equals(""))
```
Changed it throughout the class.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,187 @@
+        String name = "";
+
+        if (protocol != null && !protocol.equals(""))
+            name += protocol + "://";
```
Will work on this...

necouchman replied:
StringBuilder built.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,187 @@
+        if (paramList != null) {
+            for (String parameter : paramList) {
+                String[] paramArray = parameter.split("=");
+                qcConfig.setParameter(paramArray[0],paramArray[1]);
```
Added the limit parameter to this split, which will make sure the second `=` is part of the value.  This doesn't totally solve the issue, as you mention below...

necouchman replied:
Okay, hopefully between this tweak (adding the `limit` parameter) and the changes to address encoding, this is properly handled, now.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,187 @@
+            host = DEFAULT_URI_HOST;
+
+        if (query != null && !query.equals(""))
+            paramList = Arrays.asList(query.split("&"));
```
Yeah, good point.  Any pointers for how to handle this?

necouchman replied:
Okay, I took a stab at handling encoded parameters.

ceharris commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,204 @@
+            qcConfig.setParameter("password", password);
+
+        if (paramList != null) {
+            for (String parameter : paramList) {
```
It might make sense and would aid readability here if the parsing of the query were delegated to another method. It also seems like the kind of code that has lots of edge cases that would benefit greatly from some unit tests.

necouchman replied:
I can definitely break out the query parsing into its own method.

Never done unit tests before...another new experience for me :-).  I know there are a couple of examples in the rest of the Guacamole code that I can probably use as a starting place.

necouchman replied:
Okay, basic unit testing implemented - that wasn't too bad!

Let me know if those tests look okay...

ceharris commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -164,23 +165,23 @@ public static String getName(GuacamoleConfiguration config)
         String user = config.getParameter("username");
 
-        String name = "";
+        StringBuilder name = new StringBuilder("");
```
Just a minor nit, but `new StringBuilder()` starts out empty.

ceharris commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -131,7 +141,14 @@ public static GuacamoleConfiguration getConfiguration(String uri)
                 String[] paramArray = parameter.split("=", 2);
-                qcConfig.setParameter(paramArray[0],paramArray[1]);
+                try {
+                    qcConfig.setParameter(URLDecoder.decode(paramArray[0], "UTF-8"),
```
Might make sense to apply `URLDecoder.decode` to the entire query string once.

necouchman commented:
Okay, I think I cleaned up most of the stuff from both @mike-jumper and @ceharris for the latest round, including handling encoded parameters and some unit tests for parameter parsing.  Also tweaked a few other things on my own along the way.  Let me know what else needs to be tweaked.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/licenses/bundled/aopalliance-1.0/LICENSE:
```diff
@@ -0,0 +1,4 @@
+From http://aopalliance.sourceforge.net/:
```
Is this still bundled? Looking at the `pom.xml`, I no longer see any dependencies that are not marked as provided.

necouchman commented on extensions/guacamole-auth-quickconnect/src/licenses/LICENSE:
```diff
@@ -0,0 +1,221 @@
+terms and conditions of the following licenses.
+
+
+JUnit
```
Ah, yes, makes sense.  Fixed.

necouchman commented:
I haven't looked closely or tried it, but does anything in here need to be changed to go along with the changes for GUACAMOLE-526 that were recently merged?

necouchman commented:
Well, now I have a regression somewhere in my code where it errors out trying to open the connection initially when it is created.  So, get to go track that one down, tomorrow, too... :-/.

necouchman commented:
So, my regression appears to be some disconnect between how I've implemented the `QuickConnectREST` class and what `requestService` expects for data and returns.  The current implementation of the `create` method in `QuickConnectREST` returns a string representing the identifier of the connection that was just created:

https://github.com/apache/guacamole-client/pull/181/files#diff-6f01d367b00060b398b2a82c49ea11daR71

And the implementation of `requestService` appears to expect an object called `data` to be present:

https://github.com/apache/guacamole-client/blob/9d09f0bb0a6bed3b355572644faf374d5ca14f8b/guacamole/src/main/webapp/app/rest/services/requestService.js#L51

So, do I need to do something else in my REST class - create an object type or something like that - for this to return, or is there something else I should be doing to "wrap" this response in such a way that requestService can pass it through correctly?

necouchman commented:
> So, my regression appears to be some disconnect between how I've implemented the QuickConnectREST class and what requestService expects for data and returns.

Nevermind...figured it out.  Should be updated for the recent GUACAMOLE-526 changes.

mike-jumper commented:
I'll re-review.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java:
```diff
@@ -0,0 +1,127 @@
+
+    @Override
+    public void add(Connection connection) throws GuacamoleException {
+        connections.put(connection.getIdentifier(), connection);
```
This is not threadsafe, and may behave unexpectedly if multiple requests to add a new connection are serviced for the same user at the same time. I recommend switching from `HashMap` to `ConcurrentHashMap`.

necouchman replied:
Switched.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java:
```diff
@@ -0,0 +1,127 @@
+
+        // Create a new connection and set the parent identifier.
+        Connection connection = new SimpleConnection(name, connectionId, config);
+        connection.setParentIdentifier(ROOT_IDENTIFIER);
```
The `QuickConnectUserContext` uses `DEFAULT_ROOT_CONNECTION_GROUP`, which is not guaranteed to match `ROOT_IDENTIFIER` here. As the identifier of the root group is in fact `DEFAULT_ROOT_CONNECTION_GROUP`, that should be the value used here.

necouchman replied:
So, do I need to inject or pass in the `UserContext` object, here?  The `DEFAULT_ROOT_CONNECTION_GROUP` string is protected in the class, so I can't just use it...

necouchman replied:
Done.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectUserContext.java:
```diff
@@ -0,0 +1,123 @@
+        // Initialize the user to a SimpleUser with the provided username,
+        // no connections, and the single root group.
+        this.self = new SimpleUser(username,
+            Collections.<String>emptyList(),
```
This will only be correct when no connections exist. Once at least one connection is added, the permissions stored within this `SimpleUser` will not match up with the connections they are supposed to be able to read.

I'd recommend instead returning a new `SimpleUser` from `self()`, using `connectionDirectory.getIdentifiers()` to pull the set of readable connections.

necouchman replied:
I believe I've done this as you suggested.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectionGroup.java:
```diff
@@ -0,0 +1,112 @@
+
+        setName(name);
+        setIdentifier(identifier);
+        setType(ConnectionGroup.Type.ORGANIZATIONAL);
```
Though older code within Guacamole does this, it's bad practice to invoke overridable functions of the class being instantiated within its own constructor. If the setters belong to the superclass, `super` should be used explicitly. Some IDEs will warn of this.

https://stackoverflow.com/questions/3404301/whats-wrong-with-overridable-method-calls-in-constructors

necouchman replied:
Okay, but if the superclass doesn't have a constructor that calls these, how should I handle that?  Override the individual methods inside the class itself and then call them?  I'm looking at `AbstractConnectionGroup` and `ConnectionGroup` and `Identifiable` and not seeing a constructor that actually takes these arguments and initializes the various objects?

mike-jumper replied:
We probably should provide a constructor, but you can get around this for the time being by using `super.setName()`, `super.setIdentifier()`, etc., as that will invoke a function against the superclass which is definitely initialized at this point.

necouchman replied:
Done.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,233 @@
+
+        String username = null;
+        String password = null;
+        Map<String, String> queryParams = null;
```
There are a few cases here of the general form:

```java
String foo = null;

if (someCondition) {
    foo = somethingThatNeverReturnsNull();
}

if (foo != null) {
    doMoreThings(foo);
}
```

This is really taking what is effectively one condition and splitting it across two separate `if` blocks. The unnecessary initialization and unnecessary split can both be removed by collapsing into:

```java
if (someCondition) {
    String foo = somethingThatNeverReturnsNull();
    doMoreThings(foo);
}
```

The logic is a bit tangled otherwise.

necouchman replied:
Okay, I'll take a look at this section and try to untangle it.

necouchman replied:
Untangled.

mike-jumper commented:
Except for the above, things look generally good to me. I only have remaining nitpicks regarding the quickconnect UI itself, which looks a bit clunky IMHO:

![quickconnect-interface](https://user-images.githubusercontent.com/4632905/39779529-5392c5b8-52be-11e8-8974-711f6a41d869.png)

Perhaps there's a better way to style this?

necouchman commented:
> Perhaps there's a better way to style this?

I'm definitely open to suggestions - of the long list of things I'm not great at, GUI/graphical design is definitely toward the top of that list.  I welcome any suggestions or help with that.

necouchman commented:
@mike-jumper 
> Perhaps there's a better way to style this?

So, I tweaked the style a little bit to clean up some of the issues, but not sure it addresses all of the concerns.  I had tried to style it as much like the current interface as possible, but, as mentioned before, I'm open to suggestions - move it somewhere else, resize it, etc.  I do think it needs to be in a pretty obvious place on the screen - I wouldn't bury it in a menu - since the idea is that it allows for creating connections really quickly, so seems like minimizing mouse clicks to get there would be good.  Anyway, let me know what else should be tweaked, there.

mike-jumper commented:
Looks much much better. I'll re-review and retest. Assuming nothing new pops up, I think we're probably good to go here.

necouchman commented:
Fixed that final issue with the superclass methods.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -0,0 +1,232 @@
+                String password = userinfoMatcher.group(PASSWORD_GROUP);
+
+                if (username != null && !username.isEmpty())
+                    qcConfig.setParameter("username", username);
```
I see `parseQueryString()` automatically handling URL decoding using `URLDecoder.decode()`. Does that need to happen here, too, with the username and password?

necouchman replied:
I've added the `URLDecoder.decode()` method to this, and I also broke out the parsing of the username/password into its own method, as it was gaining the same level of complexity that the query string parser had.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/test/java/org/apache/guacamole/auth/quickconnect/utility/QCParserTest.java:
```diff
@@ -0,0 +1,49 @@
+     * Verify that the parseQueryString() method functions as designed.
+     */
+    @Test
+    public void testParseQueryString() throws UnsupportedEncodingException {
```
I suggest adding a unit test for the overall URI parsing performed by `QCParser.getConfiguration()`, which is decently complex. The parsing of query string parameters is only one piece of that pie.

necouchman replied:
Will work this up, and will also add a unit test for the new `parseUserInfo()` method.

necouchman replied:
Tests should be implemented.

necouchman commented:
Addressed the most recent couple of concerns, including adding unit tests for both the `parseUserInfo()` method and the `getConfiguration()` method.  Let me know if those tests need to process any additional corner cases.

mike-jumper commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -184,6 +184,43 @@ public static GuacamoleConfiguration getConfiguration(String uri)
+            String password = URLDecoder.decode(
+                    userinfoMatcher.group(PASSWORD_GROUP), "UTF-8");
+
+            if (username != null && !username.isEmpty())
```
Under what circumstances will `username` or `password` be null? `URLDecoder.decode()` is not documented as returning null.

necouchman replied:
Okay, moved the null check to prior to decoding it, so that should be solved.

necouchman commented on extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/utility/QCParser.java:
```diff
@@ -184,6 +184,43 @@ public static GuacamoleConfiguration getConfiguration(String uri)
+                userInfoMap.put("username", username);
+
+            if (password != null && !password.isEmpty())
+                userInfoMap.put("password", password);
```
Okay, passed the config object through to the method and just set it directly.

mike-jumper commented:
Just tested the latest from this PR, and I'm currently seeing an issue where entering a nonsensical URI like "sadjfhlaksjdhsadf" succeeds, creating a new `ssh://localhost/` connection.

necouchman commented:
> Just tested the latest from this PR, and I'm currently seeing an issue where entering a nonsensical URI like "sadjfhlaksjdhsadf" succeeds, creating a new ssh://localhost/ connection.

Hmmm...yeah, this case maybe it makes more sense to not have either a default protocol or a default host and require those items to be present at a minimum.  I'll ditch those defaults and throw some errors, instead.

necouchman commented:
Okay, got rid of the defaults, and also implemented a `QuickConnectException` class to handle error messages out of this module.  Unfortunately right now it's just throwing a standard Tomcat 500 Internal Server Error instead of passing back the error, so I have to track down why that's happening.  More to come...

necouchman commented:
Okay, @mike-jumper, I could use a little help, here, with this `QuickConnectException` class.  I've written it to extend `GuacamoleClientException` and implement `Translatable` so that it *should* provide a useful error message to the user when they enter an invalid URI.  However, instead, I'm getting a blank "ERROR" dialog, and the following in my catalina.out file:

    22-May-2018 22:20:28.708 SEVERE [http-nio-127.0.0.1-8080-exec-352] com.sun.jersey.spi.container.ContainerResponse.mapMappableContainerException The exception contained within MappableContainerException could not be mapped to a response, re-throwing to the HTTP container
    org.apache.guacamole.auth.quickconnect.QuickConnectException: URI must be absolute.
	at org.apache.guacamole.auth.quickconnect.utility.QCParser.getConfiguration(QCParser.java:83)
	at org.apache.guacamole.auth.quickconnect.rest.QuickConnectREST.create(QuickConnectREST.java:77)

I expect it to, instead, provide the proper HTTP status code (should be 400, in this case, which is set/overriden in the `GuacamoleClientException` method `getStatus()`.  What am I missing, here, that is causing it to just rethrow as HTTP 500 instead of using the assigned status code and provide the proper JSON that should include the error ("URI must be absolute.")??

mike-jumper commented:
I think this is due to the custom REST services not being intercepted in the same way that Guacamole's REST services are intercepted. There's a `RESTExceptionWrapper` thing that gets magicked around all methods that throw `GuacamoleException` and have JAX-RS annotations:

https://github.com/apache/guacamole-client/blob/1710c31fc2c0c8f56326e4175d13b21640d477ba/guacamole/src/main/java/org/apache/guacamole/rest/RESTServiceModule.java#L87-L90

For the same magic to apply to extensions, we'll need to (somehow) add the same method intercepting as part of the extension load process.

mike-jumper commented:
I think the magic will need to be added to the extension load process. That sort of introspection and run-time class modification probably should be done one time during load, not once per request.

necouchman commented:
> I think the magic will need to be added to the extension load process.

So, in the extension itself or in the extension loader?  Or both?

mike-jumper commented:
In the main extension loader. I'd think we should just somehow let the same magic apply to classes within extensions.

Probably not the sort of thing that should be the duty of the extensions, nor part of the public API.

necouchman commented:
Okay, I can't remember if the main extension loader is in the guacamole part of the project or guacamole-ext?  I'm guessing this could mean moving some of those classes over from guacamole to guacamole-ext??

necouchman commented:
Okay, I'll take a look and see if I can figure it out, open a new JIRA issue for it, and either submit a new PR or come back and beg for help because I'm in over my head :-D.

Guessing this PR needs to wait for that issue to be resolved, though, so that we avoid introducing already broken features :-/.

necouchman commented:
This appears to be working fine after implementing an `ExceptionWrapper` class, so hopefully good to go after changes for GUACAMOLE-566 get a PR and merge.

necouchman commented:
Okay, I've rebased this onto the current master after the changes in #291.  It didn't need it to function properly, but seems a little cleaner.

I had gone through and cleaned up a few things - the REST endpoint now returns JSON instead of just a string, and I touched up some comments.  Anything else need attention, here?

mike-jumper commented:
Code looks good to me. I'll retest.

@necouchman - can you update the merge base of the PR to `staging/1.0.0` now that the release process is underway?

mike-jumper commented:
Things seem to work as expected. Once the merge base is updated, I'll be happy to merge.

necouchman commented:
Merge base updated.
